### PR TITLE
Particle ids

### PR DESCRIPTION
**Pull request by Flamefire**
Requires:
- [x] #1409 
- [x] #1415 
- [x] Runtime test with Splash 1.4.0

This adds a speciesAttribute `particleId`, the traits for units and initializes the `IdProvider` in `MySimulation`.
It also implements saving and loading from HDF5 and ADIOS just like the moving window slider.

### CODE REVIEW

**Comment by bussmann** [comment]
THIS would be AWESOME!

File: src/picongpu/include/plugins/adios/ADIOSWriter.hpp [hunk 1]
```diff
@@ -543,7 +544,7 @@ class ADIOSWriter : public ILightweightPlugin
         log<picLog::INPUT_OUTPUT > ("ADIOS: value of sim_slides = %1%") %
             slides;
 
-        assert(slidesType == adiosUInt32Type.type);
+        assert( slidesType == adiosUInt32Type.type);
```
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/adios/ADIOSWriter.hpp]: one space to much

**Comment by Flamefire** [comment]
Tested and verified with ADIOS and HDF5 restart. Can be merged once the PMacc patch is merged and this is refactored.

@ax3l Please review

**Comment by ax3l** [comment]
@Flamefire pls rebase against `dev` so the PR gets shorter. can't promise I will finish the review before GTC but I will try.

**Comment by Flamefire** [comment]
Rebased

**Comment by ax3l** [comment]
ty!

File: src/picongpu/include/plugins/adios/ADIOSWriter.def [hunk 2]
```diff
@@ -74,16 +74,15 @@ namespace po = boost::program_options;
     }                                                                         \
 }
 
-#define ADIOS_CMD_EXPECT_NONZERO(_cmd)                                        \
+#define ADIOS_CMD_EXPECT_NONNULL(_cmd)                                        \
 {                                                                             \
-    int _err_code = _cmd;                                                     \
-    if (_err_code == 0)                                                       \
+    if (!(_cmd))                                                              \
```
Comment (ax3l) [inline src/picongpu/include/plugins/adios/ADIOSWriter.def]: why did you change this part? I think we explicitly wrote it that way to access the `_cmd` var only once.

Comment (Flamefire) [inline src/picongpu/include/plugins/adios/ADIOSWriter.def]: This define was never used. But I found, that one needs a define to check for Non-NULL-pointers e.g. when defining attributes. So now I can use `ADIOS_CMD_EXPECT_NONNULL( varInfo = adios_inq_var(params.fp, datasetName.c_str()) );`

Comment (ax3l) [inline src/picongpu/include/plugins/adios/ADIOSWriter.def]: great!

File: src/picongpu/include/plugins/hdf5/HDF5Writer.hpp [hunk 3]
```diff
@@ -221,6 +221,15 @@ class HDF5Writer : public ISimulationPlugin
         ForEach<FileCheckpointParticles, LoadSpecies<bmpl::_1> > forEachLoadSpecies;
         forEachLoadSpecies(params, restartChunkSize);
 
+        IdProvider<simDim>::State idProvState;
+        ReadNDScalars<uint64_t, uint64_t>()(mThreadParams,
+                "picongpu/idProvider/startId", &idProvState.startId,
+                "maxNumProc", &idProvState.maxNumProc);
+        ReadNDScalars<uint64_t>()(mThreadParams,
+                "picongpu/idProvider/nextId", &idProvState.nextId);
+        log<picLog::INPUT_OUTPUT > ("Setting id on current rank: %1%") % idProvState.nextId;
```
Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/HDF5Writer.hpp]: be more explicit, let us write "Setting next id on ..."

File: src/picongpu/include/plugins/adios/NDScalars.hpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+namespace picongpu {
+namespace adios {
+
+/* Functors for reading and writing ND scalar fields with N=simDim
+ * In the current implementation each process (of the ND grid) reads/writes 1 scalar value
+ * Optionally the processes can also read/write an attribute for this dataset by using a non-empty attrName
+ */
+
+template<typename T_Skalar, typename T_Attribute = uint64_t>
```
Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: typo: `T_Scalar`

File: src/picongpu/include/plugins/adios/NDScalars.hpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+struct WriteNDScalars
+{
+    WriteNDScalars(const std::string& name, const std::string& attrName = ""):name(name), attrName(attrName){}
```
Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: pls put the initialization on the next line

File: src/picongpu/include/plugins/adios/NDScalars.hpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        log<picLog::INPUT_OUTPUT> ("ADIOS: write %1%D scalars: %2%") % simDim % name;
+
+        ADIOS_CMD( adios_write_byid(params.adiosFileHandle, varId, &value) );
+    }
+private:
+    const std::string name, attrName;
+    int64_t varId;
+};
+
+template<typename T_Skalar, typename T_Attribute = uint64_t>
```
Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: `T_Scalar` typo

File: src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    /**
+     * Read an attribute from an open ADIOS file, check that its type is correct and return it
+     *
+     * @param fp       Open ADIOS file handle
+     * @param basePath Path where the attribute is located in the file (with or w/o trailing slash)
+     * @param attrName Name of the attribute. Used for status output and concatenated with basePath
+     * @retval Attribute value
+     */
+    template<typename T_Attribute>
+    T_Attribute readAttribute(ADIOS_FILE* fp, const std::string& basePath, const std::string& attrName)
```
Comment (ax3l) [inline src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp]: don't we just the the same thing while reading `sim_slides` but way less verbose?

Comment (ax3l) [inline src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp]: otherwise: let's read `sim_slides` the same way so we have the same check.

btw: can you move sim_slide to the new `picongpu/...` dir, too?

File: src/picongpu/include/plugins/hdf5/NDScalars.hpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace picongpu {
+namespace hdf5 {
+
+/* Functors for reading and writing ND scalar fields with N=simDim
+ * In the current implementation each process (of the ND grid) reads/writes 1 scalar value
+ * Optionally the processes can also read/write an attribute for this dataset
+ */
+
+template<typename T_Skalar, typename T_Attribute = uint64_t>
```
Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/NDScalars.hpp]: Typo `T_Scalar`

File: src/picongpu/include/plugins/hdf5/NDScalars.hpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+#include "pmacc_types.hpp"
+#include "plugins/hdf5/HDF5Writer.def"
+#include "traits/PICToSplash.hpp"
+#include "Environment.hpp"
+
+namespace picongpu {
+namespace hdf5 {
+
+/* Functors for reading and writing ND scalar fields with N=simDim
```
Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/NDScalars.hpp]: just a side note: even if `simDim` is 3 it does not mean one needs to do the domain decomposition in 3 dimensions. In the context of ID providers this is a bit irritating, but ... ook.

File: src/picongpu/include/plugins/hdf5/NDScalars.hpp [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+ * In the current implementation each process (of the ND grid) reads/writes 1 scalar value
```
Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/NDScalars.hpp]: I would write ND domain decomposition, grid is reserved in our speech for actual cells

File: src/picongpu/include/simulation_defines/unitless/speciesAttributes.unitless [hunk 11]
```diff
@@ -194,6 +195,25 @@ struct UnitDimension<weighting >
     }
 };
 
+template<>
+struct Unit<particleId>
+{
+    /* zero-sized vector indicating unitless flag for hdf5 and adios output */
```
Comment (ax3l) [inline src/picongpu/include/simulation_defines/unitless/speciesAttributes.unitless]: that is right now changing in #1428 .

please make it directly
```C++
    /* unitless and not scaled by a factor: by convention 1.0 */
    static std::vector<double> get()
    {
        std::vector<double> unit( 1, 1.0 );
        return unit;
    }
```

File: src/picongpu/include/traits/AdiosToPIC.tpp [hunk 12]
```diff
@@ -44,6 +44,12 @@ namespace traits
         typedef unsigned int type;
     };
 
+    template<>
+    struct AdiosToPIC<adios_unsigned_long>
+    {
+        typedef uint64_t type;
```
Comment (ax3l) [inline src/picongpu/include/traits/AdiosToPIC.tpp]: I think `unsigned long` would be better (see above)

Comment (ax3l) [inline src/picongpu/include/traits/AdiosToPIC.tpp]: pls add `adios_long` for completeness

File: src/picongpu/include/traits/PICToAdios.tpp [hunk 13]
```diff
@@ -60,6 +60,15 @@ namespace traits
         type(adios_unsigned_long) {}
     };
 
+    template<>
+    struct PICToAdios<uint64_cu>
```
Comment (ax3l) [inline src/picongpu/include/traits/PICToAdios.tpp]: hm... necessary? can't you cast to `unit64_t` before?

Comment (ax3l) [inline src/picongpu/include/traits/PICToAdios.tpp]: above: I think our current implementation is not ideal, we should do
```C++
/** Trait for unsigned long */
     template<>
     struct PICToAdios<unsigned long>
     {
         ADIOS_DATATYPES type;
         
         PICToAdios() :
         type(adios_unsigned_long) {}
     };
```

since adios does not identify 64 and 32bit directly for ints.

File: src/picongpu/include/traits/PICToSplash.tpp [hunk 14]
```diff
@@ -57,6 +57,18 @@ namespace traits
...
     };
 
+    template<>
+    struct PICToSplash<uint64_t>
+    {
+        typedef splash::ColTypeUInt64 type;
+    };
+
+    template<>
+    struct PICToSplash<uint64_cu>
```
Comment (ax3l) [inline src/picongpu/include/traits/PICToSplash.tpp]: same here, better use `uint64_t` for writing if possible.

File: src/picongpu/include/plugins/adios/NDScalars.hpp [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+#include <stdexcept>
+
+namespace picongpu {
+namespace adios {
+
+/* Functors for reading and writing ND scalar fields with N=simDim
+ * In the current implementation each process (of the ND grid) reads/writes 1 scalar value
+ * Optionally the processes can also read/write an attribute for this dataset by using a non-empty attrName
+ */
+
```
Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: no space between doxygen und class impl pls :)

Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: add `/**` to make it doxygen-ish

Comment (ax3l) [inline src/picongpu/include/plugins/adios/NDScalars.hpp]: pls add `@tparam descriptors`

**Comment by ax3l** [comment]
comments are inline, looks good to me if the runtime tests were all stable.

@psychocoderHPC please have a look about the new `traits`, too (adios exposes no `uint64_t` just `int` and `long` which I think we should specialize for).

**Comment by psychocoderHPC** [comment]
Please rebase your branch because #1427 changed the `libSplash` version to `1.4.0` and start a runtime test to verify your results.
Maybe this not effects your branch.

File: src/picongpu/include/traits/PICToAdios.tpp (hunk 13 above)
Comment (psychocoderHPC) [inline src/picongpu/include/traits/PICToAdios.tpp]: We can't cast our data befor we write because it is a generic method.
But as wrote in our chat session we can test if it is possible to define `struct PICToAdios<unsigned long int>` and `struct PICToAdios<unsigned long long int>`.

File: src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp (hunk 7 above)
Comment (Flamefire) [inline src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp]: Yeah I have seen quite some duplicated code and before adding another instance of the code I generalized it into a functor.
Moving is possible, but please open an issue and let's do this in another PR. They are unaffected by this and can be done separately AFTER this one.

File: src/picongpu/include/plugins/hdf5/NDScalars.hpp (hunk 9 above)
Comment (Flamefire) [inline src/picongpu/include/plugins/hdf5/NDScalars.hpp]: If simDim is 3 than the processes are organized in a 3D grid (or however you name them). And per definition (of this functor) every process writes 1 value.

File: src/picongpu/include/plugins/hdf5/NDScalars.hpp (hunk 10 above)
Comment (Flamefire) [inline src/picongpu/include/plugins/hdf5/NDScalars.hpp]: This is less related to the domain but the the process organization. So I'd rather write `ND grid of processes`

File: src/picongpu/include/traits/AdiosToPIC.tpp (hunk 12 above)
Comment (Flamefire) [inline src/picongpu/include/traits/AdiosToPIC.tpp]: We would not be the only ones doing that: https://github.com/Kitware/VTK/blob/master/IO/ADIOS/ADIOSScalar.cxx#L64, http://fossies.org/linux/VTK/IO/ADIOS/ADIOSUtilities.cxx

File: src/picongpu/include/traits/PICToSplash.tpp (hunk 14 above)
Comment (Flamefire) [inline src/picongpu/include/traits/PICToSplash.tpp]: See above, it is required unfortunately

File: src/picongpu/include/traits/AdiosToPIC.tpp (hunk 12 above)
Comment (psychocoderHPC) [inline src/picongpu/include/traits/AdiosToPIC.tpp]: Offline discussed with many emotions :-)
Adios defines `adios_unsigned_long` as `unsigned long long` see [here](https://github.com/ornladios/ADIOS/blob/master/src/core/common_read.c#L1362-L1363)

We should do the same even if `unsigned long long` is not defined in `C++03` but in `C++11`.
`unsigned long long` is defined in `C99` and Adios is C code.

`unsigned long long` in C++11 is defined as `at least 64bit` see [here](http://de.cppreference.com/w/cpp/language/types)

**Comment by Flamefire** [comment]
Code is rebased and all suggestions so far introduced except the ADIOS types thing where there was some discussion. I added the missing signed versions but used the standard fixed-width typedefs and a conditional for the `uint64_cu` so it is only defined when its not equal to `uint64_cu`. Reasoning:
- ADIOS may **use** `unsigned long long` but **defines** `adios_unsigned_long` to refer to an 8-byte type: https://github.com/ornladios/ADIOS/blob/37d0b2b44899e1c2bb1e538f2b949e39bcde68a/src/public/adios_types.h#L16-L26. This file belongs to the public interface and should be the reference. Not some (probably faulty) internal function
- ADIOS also enforces the width to be 8 bytes: https://github.com/ornladios/ADIOS/blob/37d0b2b44899e1c2bb1e538f2b949e39bcde68a1/src/core/adios_bp_v1.c#L2387-L2389
- So passing anything that is not an **exactly** 8 byte type will cause an error: https://github.com/ornladios/ADIOS/blob/37d0b2b44899e1c2bb1e538f2b949e39bcde68a1/src/core/common_adios.c#L768

I adapted the other trait specializations too, to have a consistent and correct implementation.

**Comment by psychocoderHPC** [comment]
@Flamefire Thanks for the links to the Adios implementation.

File: src/picongpu/include/traits/PICToAdios.tpp [hunk 16]
```diff
@@ -32,7 +34,7 @@ namespace traits
 {
     /** Trait for int */
     template<>
-    struct PICToAdios<int>
+    struct PICToAdios<int32_t>
```
Comment (psychocoderHPC) [inline src/picongpu/include/traits/PICToAdios.tpp]: if you change this you should also change the definition in `PICToSplash`

File: src/picongpu/include/traits/PICToAdios.tpp [hunk 17]
```diff
@@ -32,7 +34,7 @@ namespace traits
 {
     /** Trait for int */
     template<>
-    struct PICToAdios<int>
```
Comment (psychocoderHPC) [inline src/picongpu/include/traits/PICToAdios.tpp]: If we remove this with fixed size types than we should add an description at the begin of the files that it is not allowed to define native C/C++ includes types e.g., `unsigned long`, `int` ,... because they can conflict with the types `int32_t`, `uint64_t`, ...

File: src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp (hunk 7 above)
Comment (ax3l) [inline src/picongpu/include/plugins/adios/restart/ReadAttribute.hpp]: ok for me!

**Comment by ax3l** [comment]
ok for me. The only thing is: I would avoid `uint64_cu` alltogether in the code but we can do it with the enable_if for now.

**Comment by ax3l** [comment]
@psychocoderHPC ok from my side, pls merge if you are ok, too.

File: src/picongpu/include/traits/PICToAdios.tpp (hunk 17 above)
Comment (Flamefire) [inline src/picongpu/include/traits/PICToAdios.tpp]: Using them would simply not compile if someone was to forget this. So I see no need for that description

File: src/picongpu/include/traits/PICToAdios.tpp (hunk 16 above)
Comment (Flamefire) [inline src/picongpu/include/traits/PICToAdios.tpp]: Changed that anyway so it is now consistent

File: src/picongpu/include/traits/PICToAdios.tpp (hunk 17 above)
Comment (psychocoderHPC) [inline src/picongpu/include/traits/PICToAdios.tpp]: I mean not a comment for the user but a comment for developer that we not add types like int or unsigned long in a half year because we forget that it can conflict with uint32_t.
This double definition can compile but if a user switch to a a tegra or similar maybe it crash.

Comment (Flamefire) [inline src/picongpu/include/traits/PICToAdios.tpp]: Well it has a) to be merged and b) pass the tests on the CI suite. So I don't think it would compile. (How could it, if it conflicts?) And if it even crashes then it is a bug introduced by the developer and undetected by the reviewer.
I just think that a comment like this is nowhere else and if we start adding comments with coding guidelines to the code there will be quite some code bloat with something that should be obvious or (if at all) documented in the general coding guidelines.

**Comment by Flamefire** [comment]
I don't see that we can avoid `uint64_cu` as the cuda API is defined with that type. Especially the atomics. So with the conditional here we are at least flexible and if cuda was to use `uint64_t` in the future, than this code does not need to change.

File: src/picongpu/include/traits/PICToAdios.tpp (hunk 17 above)
Comment (ax3l) [inline src/picongpu/include/traits/PICToAdios.tpp]: yes the problem is more if you go a system like tk1 which has a 32bit host, the compile suite will not cross-compile to it for testing.

nevertheless, I think it's fine as it is @psychocoderHPC

**Comment by ax3l** [comment]
Why do you use it in `IDProvider` again? You could also use a `HostDeviceBuffer<uint64_t, 1>` buffer and be set, no?

**Comment by Flamefire** [comment]
The `atomicAdd` is defined on a `uint64_cu`. I could probably make the idProvider itself return an uint64_t using an implicit or explicit conversion. Might work as well. Just wanted to avoid that conversion as we do have that type

**Comment by Flamefire** [comment]
@psychocoderHPC I removed the uint64_cu from the interface. So I consider this ready for merge

**Comment by psychocoderHPC** [comment]
Thanks for this nice new feature.

**Note:** Next time we should split it up into two or three pull requests. But nobody is still using this feature therefore it can be merged to `dev`.

**Comment by ax3l** [comment]
@Flamefire did you do a runtime test after your last changes? :)

When rebasing against the new changes that got into dev (first view: this branch) I get a `boost::traits` runtime error
```
boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::io::too_many_args> >
boost::too_many_args: format-string referred to less arguments than were passed
```

I am just guessing it might be this PR since it occurs on enabled HDF5 output since the last rebase.

File: src/picongpu/include/plugins/hdf5/HDF5Writer.hpp [hunk 18]
```diff
@@ -452,6 +462,14 @@ class HDF5Writer : public ISimulationPlugin
         }
         log<picLog::INPUT_OUTPUT > ("HDF5: ( end ) writing particle species.");
 
+        PMACC_AUTO(idProviderState, IdProvider<simDim>::getState());
+        log<picLog::INPUT_OUTPUT>("HDF5: Writing IdProvider state (StartId: %1%, NextId: %1%, maxNumProc: %2%)")
```
Comment (ax3l) [inline src/picongpu/include/plugins/hdf5/HDF5Writer.hpp]: found the RT error :)

**Comment by ax3l** [comment]
pushed the fix to #1440 ;)

**Comment by Flamefire** [comment]
I actually did test it, but I guess this bug just shows when the log option is enabled. Thanks for fixing :+1:

**Comment by ax3l** [comment]
Yes that's what probably happened, no prob :)
